### ISSUE TITLE
constant_folding ignores epsilon

### ISSUE THREAD

**Issue by szutenberg** [tensorflow/tensorflow#52045]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes. [https://gist.github.com/szutenberg/76b35f503195e8c7b2f1d6d764ee1a6f]
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 18.04
- TensorFlow installed from (source or binary): binary (pip)
- TensorFlow version (use command below): v2.6.0-rc2-32-g919f693420e 2.6.0 (CPU version)
- Python version: 3.6.9

When running the script https://gist.github.com/szutenberg/76b35f503195e8c7b2f1d6d764ee1a6f I'm getting:
```
=== EPSILON =  1e-09  dtype =  <dtype: 'float32'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-0.       -0.       15.942385       inf]
=== EPSILON =  1e-07  dtype =  <dtype: 'float32'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-1.1920928e-07 -1.1920928e-07  1.5249238e+01  1.5942385e+01]
=== EPSILON =  1e-09  dtype =  <dtype: 'float64'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-1.00000008e-09  0.00000000e+00  1.61081453e+01  2.07232658e+01]
=== EPSILON =  1e-07  dtype =  <dtype: 'float64'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-9.99999951e-08 -9.89999951e-08  1.54249485e+01  1.61180957e+01]
```

Add( Sub(1, logits), epsilon) is simplified by grappler (constant folding) to Sub(1+epsilon, logits)

User doesn't expect inf because epsilon was used for this purpose. It's expected to see:
```
=== EPSILON =  1e-09  dtype =  <dtype: 'float32'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-0.        -0.        15.9340315 20.723267 ]
=== EPSILON =  1e-07  dtype =  <dtype: 'float32'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-1.1920928e-07 -1.1920928e-07  1.5333239e+01  1.6118095e+01]
=== EPSILON =  1e-09  dtype =  <dtype: 'float64'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-1.00000008e-09  0.00000000e+00  1.61081453e+01  2.07232658e+01]
=== EPSILON =  1e-07  dtype =  <dtype: 'float64'>
in  =  [0.000000e+00 1.000000e-09 9.999999e-01 1.000000e+00]
out =  [-9.99999951e-08 -9.89999951e-08  1.54249485e+01  1.61180957e+01]
```

![image](https://user-images.githubusercontent.com/37601244/133803576-2b598fef-247a-4794-9df4-3e6208ec8b5c.png)
As we can see, it's impossible to represent 1+1e-9 in float32. It's rounded to 1.0 which causes inf (which very likely will break the training with NaN).

This case is extracted from model using hard negative mining. The input comes from sigmoid.

**[Contributing](https://www.tensorflow.org/community/contribute)**

- Do you want to contribute a PR? (yes/no): Yes
- Briefly describe your candidate solution(if contributing):
constant folding should not simplify A+B = A when B != 0

**Standalone code to reproduce the issue**
https://gist.github.com/szutenberg/76b35f503195e8c7b2f1d6d764ee1a6f

Does it make sense to prepare patch resolving this issue?

**Comment by mohantym** [comment]
Hi, @Saduf2019 ,Could you please at this issue ? Replicating in [2.5](https://colab.research.google.com/gist/mohantym/bb53196c75e3ab96330dbd93c480c0a0/github_52045.ipynb#scrollTo=oS1QZpXWkL_Z),[2.6 ](https://colab.research.google.com/gist/mohantym/5daa1fbb63634d089a455d5a7da86bc9/github_52045.ipynb#scrollTo=oS1QZpXWkL_Z)and [2.7](https://colab.research.google.com/gist/mohantym/49df0055e3b3061a61e03fb2fa0451c4/github_52045.ipynb#scrollTo=mnaSBNXHWSf5).

**Comment by mohantym** [comment]
Hi @szutenberg !
1e-07 would be a safer epsilon value compatible to TF 2.x versions.  You can also use tf.keras.backend.epsilon for not using epsilon values explicitly. Attached [gist](https://colab.sandbox.google.com/gist/mohantym/532227499de57f8edb3d2a7a8cafd87e/untitled99.ipynb#scrollTo=CPdYvo3G49Qb) for 
[Reference.]( https://stackoverflow.com/a/66332679/11530462)
Thank you!

**Comment by szutenberg** [comment]
Hi @mohantym ,
Sure. Epsilon 1e-07 works but my point is that grappler's constant folding doesn't work correctly.

**Comment by mohantym** [comment]
Ok @szutenberg ! 
Thanks for the update!

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
